### ISSUE TITLE
some questions on enabling a specific algorithm & celery workers constant usage optimization

### ISSUE THREAD

**Issue by flamby** [ARM-software/mango#17]
Hello,

First of all, kudos for your lib, it's by far the more convenient in the subject I've tested so far, particularly when one wants to search in // using multiple nodes.

**Hallucination / Clustering mode**
Having read the mango slides PDF, I've difficulties to match in the `Tuner` _conf dict_ the **hallucination** & **clustering** approaches depicted in Mango Experiments.

Reading the `1D_Mango_Different_Parallelization.ipynb` or other notebooks did not clarify that for me.

Is it just a matter of batch > 1 to enable what you call the "hallucination" mode? What about the clustering mode?
I think the _README.md_ could be improved in this area. Are those algorithms compatible to celery workers usage?

**Celery workers constant usage**
So far, I've seen that my celery workers aren't used constantly, which makes sense since some experiment results might guide next experiments.

I used the `simple_celery.py` example approach, tried different values for `n_jobs` of the `@scheduler.custom` decorator, as well as several `batch_size` and `num_iteration` but none of these approaches avoid the lack of use of some workers for a long time.
Is there any tuning recommendations to optimize the celery worker usage?

Each task takes around 30 to 60 seconds to complete, as I'm not searching for hyper-parameters, but instead using mango to search for parameters needed for features engineering, and features generation takes time with my use-case.
Can this long celery task duration be part of my issue?

BTW, why not using the `@scheduler.celery` decorator in your celery examples?

Thanks, and keep the good work!

**Comment by sandeep-iitr** [comment]
Thanks for asking the question. We will update the clarification in the Readme file to clarify some of these questions better.

**Hallucination / Clustering mode**
1. By default, when _batch_size_ >1, the parallel mode is hallucination.
2. you can specify different parallel modes with `conf_dict['parallel_strategy']` value. Set it to `'penalty'` for hallucination approach, and use the value of `'clustering'` for the clustering algorithm.

Yes, these algorithms are compatible with celery workers or any parallel scheduling frameworks. Basically, these algorithms will give the objective function a batch of values that can be evaluated in parallel using any framework. The specific choice of objective function evaluation is left to the user. We do provide examples for celery. The implemented parallel scheduler decorator (`@scheduler.parallel(n_jobs=2)`) uses the joblib, and has a structure that doesn't expose batch values directly. A simple celery example that exposes all the batch values to userobjective (`objectiveKNNCelery(args_list)`) is available [here](https://github.com/ARM-software/mango/blob/master/examples/KNN_Celery.ipynb). The batch is passed as an `args_list` parameter.


**Celery workers constant usage**
The number of parallel evaluations at any instance running is equal to the batch of values. After evaluating a batch, the next batch of values is selected, the selection of batch is done in serial. The selection process is much faster as compared to the typical ML classifier training time, so it is generally not directly visible for longer training times. 

You can make the size of the batch equal to the number of celery workers, and follow the template from [Eg1](https://github.com/ARM-software/mango/blob/master/examples/KNN_Celery.ipynb), [Eg2](https://github.com/ARM-software/mango/blob/master/examples/Prophet_Celery.ipynb) to directly schedule the batch on celery workers. 

_Each task takes around 30 to 60 seconds to complete, as I'm not searching for hyper-parameters, but instead using mango to search for parameters needed for features engineering, and features generation takes time with my use-case.
Can this long celery task duration be part of my issue?_
In this, my recommendation is to use the clustering algorithm which is shown to converge faster for most of the examples with better batch values, and use batch size equal to the number of celery workers, so each batch is running. Follow the simple example templates mentioned earlier to directly use your celery workers by scheduling your batches.

_BTW, why not using the @scheduler.celery decorator in your celery examples?_
This is a good suggestion, we will look into it and see how to provide this for simple settings. Generally, we also use celery to schedule tasks on the Kubernetes cluster, so we have kept that as a choice to users to define their scheduler, but providing a simple decorator can help the users. 

If you face any more problems/questions, let us know.

**Comment by flamby** [comment]
Hello @sandeep-iitr 

Thanks a lot for these explanations. I know what to do now and I'm eager to see the results :-)

Thanks again.
